pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -166,17 +169,23 @@ async def _prepare_messages(
                 ctx_messages.used = True
 
         parts: list[_messages.ModelRequestPart] = []
-        if message_history:
-            # Shallow copy messages
-            messages.extend(message_history)
-            # Reevaluate any dynamic system prompt parts
-            await self._reevaluate_dynamic_prompts(messages, run_context)
+        if self.instructions or self.instructions_functions:
+            instructions = await self._instructions(run_context)
+            if message_history:
+                messages.extend(message_history)
         else:
-            parts.extend(await self._sys_parts(run_context))
+            instructions = None
+            if message_history:
+                # Shallow copy messages
+                messages.extend(message_history)
+                # Reevaluate any dynamic system prompt parts
+                await self._reevaluate_dynamic_prompts(messages, run_context)
```

Comment:
We should do this whether or not there are instructions; I think it will allow us to remove the logic fork above contingent on the presence of instructions/instruction functions.